File: 023_merge_k_sorted_lists/Solution.java
import java.util.PriorityQueue;

class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;

        ListNode dummyHead = new ListNode(0);
        ListNode curNode = dummyHead;
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);

        for (ListNode list : lists) {
            ListNode tmp = list;
            while (tmp != null) {
                pq.add(tmp);
                tmp = tmp.next;
            }
        }

        while (!pq.isEmpty()) {
            curNode.next = pq.poll();
            curNode = curNode.next;
        }
        curNode.next = null; // 防止出现环
        
        return dummyHead.next;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        //ListNode l1 = new ListNode(1);
        //l1.next = new ListNode(4);
        //l1.next.next = new ListNode(5);
        //
        //ListNode l2 = new ListNode(1);
        //l2.next = new ListNode(3);
        //l2.next.next = new ListNode(4);
        //
        //ListNode l3 = new ListNode(2);
        //l3.next = new ListNode(6);
        //ListNode[] lists = new ListNode[]{l1, l2, l3};

        ListNode l1 = new ListNode(-2);
        l1.next = new ListNode(-1);
        l1.next.next = new ListNode(-1);
        l1.next.next.next = new ListNode(-1);
        ListNode[] lists = new ListNode[]{l1, null};

        ListNode head = solution.mergeKLists(lists);
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.next;
        }
    }
}

File: 023_merge_k_sorted_lists/Solution.go
package main

import (
	"container/heap"
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

type PriorityQueue []*ListNode

func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].Val < pq[j].Val
}

func (pq *PriorityQueue) Swap(i, j int) {
	(*pq)[i].Val, (*pq)[j].Val = (*pq)[j].Val, (*pq)[i].Val
}

func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*ListNode))
}

func (pq *PriorityQueue) Pop() interface{} {
	result := (*pq)[len(*pq)-1]
	*pq = (*pq)[:len(*pq)-1]
	return result
}

func mergeKLists(lists []*ListNode) *ListNode {
	if lists == nil || len(lists) == 0 {
		return nil
	}

	dummyHead := &ListNode{0, nil}
	curNode := dummyHead
	var pq PriorityQueue

	for _, list := range lists {
		tmp := list
		for tmp != nil {
			heap.Push(&pq, tmp)
			tmp = tmp.Next
		}
	}

	for len(pq) != 0 {
		curNode.Next = heap.Pop(&pq).(*ListNode)
		curNode = curNode.Next
	}
	curNode.Next = nil

	return dummyHead.Next
}

func main() {
	l1 := &ListNode{1, nil}
	l1.Next = &ListNode{4, nil}
	l1.Next.Next = &ListNode{5, nil}

	l2 := &ListNode{1, nil}
	l2.Next = &ListNode{3, nil}
	l2.Next.Next = &ListNode{4, nil}

	l3 := &ListNode{2, nil}
	l3.Next = &ListNode{6, nil}

	lists := []*ListNode{l1, l2, l3}

	result := mergeKLists(lists)
	for result != nil {
		fmt.Printf("%d ", result.Val)
		result = result.Next
	}
}

